From da2b388116314a7df28ecae1f59677fb6deab646 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Tue, 3 Feb 2015 13:48:19 +0530
Subject: [PATCH] [TYPE_C]: Add support type-c in power supply framework

This patch add support for Type-C related detections and charging (charger
type, cable type, device type, etc) in power supply framework.

Change-Id: Ie3c3dd16dd171dbb7667829c3a17a103049fdc59
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-20288
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
Signed-off-by: R, Kannappan <r.kannappan@intel.com>
---
 drivers/extcon/extcon-class.c        | 1 +
 drivers/power/power_supply_charger.c | 6 ++++++
 drivers/power/power_supply_charger.h | 5 +++--
 drivers/power/power_supply_sysfs.c   | 2 +-
 include/linux/extcon.h               | 1 +
 include/linux/power_supply.h         | 5 ++++-
 6 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/drivers/extcon/extcon-class.c b/drivers/extcon/extcon-class.c
index f488982..af03c9c 100644
--- a/drivers/extcon/extcon-class.c
+++ b/drivers/extcon/extcon-class.c
@@ -53,6 +53,7 @@ const char extcon_cable_name[][CABLE_NAME_MAX + 1] = {
 	[EXTCON_CDP]		= "CHARGER_USB_CDP",
 	[EXTCON_ACA]		= "CHARGER_USB_ACA",
 	[EXTCON_AC]		= "CHARGER_AC",
+	[EXTCON_TYPEC]		= "USB_TYPEC",
 	[EXTCON_HDMI]		= "HDMI",
 	[EXTCON_MHL]		= "MHL",
 	[EXTCON_DVI]		= "DVI",
diff --git a/drivers/power/power_supply_charger.c b/drivers/power/power_supply_charger.c
index 141424f..c5956e8 100644
--- a/drivers/power/power_supply_charger.c
+++ b/drivers/power/power_supply_charger.c
@@ -70,6 +70,10 @@ static struct charger_cable cable_list[] = {
 	 .psy_cable_type = POWER_SUPPLY_CHARGER_TYPE_AC,
 	 .extcon_cable_type = EXTCON_AC,
 	 },
+	{
+	 .psy_cable_type = POWER_SUPPLY_CHARGER_TYPE_USB_TYPEC,
+	 .extcon_cable_type = EXTCON_TYPEC,
+	},
 };
 
 static int get_supplied_by_list(struct power_supply *psy,
@@ -103,6 +107,8 @@ struct charger_cable *get_cable(unsigned long usb_chrgr_type)
 		return &cable_list[4];
 	case POWER_SUPPLY_CHARGER_TYPE_AC:
 		return &cable_list[6];
+	case POWER_SUPPLY_CHARGER_TYPE_USB_TYPEC:
+		return &cable_list[7];
 	case POWER_SUPPLY_CHARGER_TYPE_SE1:
 		return &cable_list[5];
 	}
diff --git a/drivers/power/power_supply_charger.h b/drivers/power/power_supply_charger.h
index 3dfb0df..7790edf 100644
--- a/drivers/power/power_supply_charger.h
+++ b/drivers/power/power_supply_charger.h
@@ -183,8 +183,9 @@ static inline int get_ps_int_property(struct power_supply *psy,
 #define IS_BATTERY(psy) (psy->type == POWER_SUPPLY_TYPE_BATTERY)
 #define IS_CHARGER(psy) (psy->type == POWER_SUPPLY_TYPE_USB ||\
 				psy->type == POWER_SUPPLY_TYPE_USB_CDP || \
-			psy->type == POWER_SUPPLY_TYPE_USB_DCP ||\
-			psy->type == POWER_SUPPLY_TYPE_USB_ACA)
+			psy->type == POWER_SUPPLY_TYPE_USB_DCP || \
+			psy->type == POWER_SUPPLY_TYPE_USB_ACA || \
+			psy->type == POWER_SUPPLY_TYPE_USB_TYPEC)
 #define IS_ONLINE(psy) \
 		(get_ps_int_property(psy, POWER_SUPPLY_PROP_ONLINE) == 1)
 #define IS_PRESENT(psy) \
diff --git a/drivers/power/power_supply_sysfs.c b/drivers/power/power_supply_sysfs.c
index 77a5992..5ebd412 100644
--- a/drivers/power/power_supply_sysfs.c
+++ b/drivers/power/power_supply_sysfs.c
@@ -45,7 +45,7 @@ static ssize_t power_supply_show_property(struct device *dev,
 					  char *buf) {
 	static char *type_text[] = {
 		"Unknown", "Battery", "UPS", "Mains", "USB", "USB",
-		"USB_DCP", "USB_CDP", "USB_ACA"
+		"USB_DCP", "USB_CDP", "USB_ACA", "USB_HOST", "USB_TYPEC"
 	};
 	static char *status_text[] = {
 		"Unknown", "Charging", "Discharging", "Not charging", "Full"
diff --git a/include/linux/extcon.h b/include/linux/extcon.h
index 89d7dd0..f618658 100644
--- a/include/linux/extcon.h
+++ b/include/linux/extcon.h
@@ -61,6 +61,7 @@ enum extcon_cable_name {
 	EXTCON_CDP,
 	EXTCON_ACA,
 	EXTCON_AC,
+	EXTCON_TYPEC,
 	EXTCON_HDMI,
 	EXTCON_MHL,
 	EXTCON_DVI,
diff --git a/include/linux/power_supply.h b/include/linux/power_supply.h
index 7798a6b..b4ce7b0 100644
--- a/include/linux/power_supply.h
+++ b/include/linux/power_supply.h
@@ -184,6 +184,7 @@ enum power_supply_type {
 	POWER_SUPPLY_TYPE_USB_CDP,	/* Charging Downstream Port */
 	POWER_SUPPLY_TYPE_USB_ACA,	/* Accessory Charger Adapters */
 	POWER_SUPPLY_TYPE_USB_HOST,	/* To support OTG devices */
+	POWER_SUPPLY_TYPE_USB_TYPEC,    /* To support TYPEC devices */
 };
 
 enum power_supply_charger_event {
@@ -214,6 +215,7 @@ enum power_supply_charger_cable_type {
 	POWER_SUPPLY_CHARGER_TYPE_SE1 = 1 << 9,
 	POWER_SUPPLY_CHARGER_TYPE_MHL = 1 << 10,
 	POWER_SUPPLY_CHARGER_TYPE_B_DEVICE = 1 << 11,
+	POWER_SUPPLY_CHARGER_TYPE_USB_TYPEC = 1 << 12,
 };
 
 enum power_supply_notifier_events {
@@ -235,7 +237,8 @@ struct power_supply_cable_props {
 	POWER_SUPPLY_CHARGER_TYPE_USB_CDP | \
 	POWER_SUPPLY_CHARGER_TYPE_USB_ACA | \
 	POWER_SUPPLY_CHARGER_TYPE_ACA_DOCK| \
-	POWER_SUPPLY_CHARGER_TYPE_SE1)
+	POWER_SUPPLY_CHARGER_TYPE_SE1 |	\
+	POWER_SUPPLY_CHARGER_TYPE_USB_TYPEC)
 
 union power_supply_propval {
 	int intval;
-- 
1.9.1

